pilot/pkg/model/push_context.go

```diff
@@ -419,6 +386,38 @@ func (ps *PushContext) VirtualServices(proxy *Proxy, gateways map[string]bool) [
 	return out
 }
 
+// GetSidecarScope returns a sidecar rule applicable to the config
+// namespace associated with the proxy
+func (ps *PushContext) GetSidecarScope(proxy *Proxy, proxyInstances []*ServiceInstance) *SidecarScope {
+
+	var workloadLabels LabelsCollection
+	for _, w := range proxyInstances {
+		workloadLabels = append(workloadLabels, w.Labels)
+	}
+
+	if sidecars, ok := ps.sidecarsByNamespace[proxy.ConfigNamespace]; ok {
+		// TODO: logic to merge multiple sidecar resources
+		// Currently we assume that there will be only one sidecar config for a namespace.
+		for _, wrapper := range sidecars {
+			if wrapper.Config != nil {
+				sidecar := wrapper.Config.Spec.(*networking.Sidecar)
+				// if there is no workload selector, the config applies to all workloads
+				// if there is a workload selector, check for matching workload labels
+				if sidecar.GetWorkloadSelector() != nil {
+					workloadSelector := Labels(sidecar.GetWorkloadSelector().GetLabels())
+					if !workloadLabels.IsSupersetOf(workloadSelector) {
+						continue
+					}
+					return wrapper
+				}
+			}
+			return wrapper
+		}
+	}
+
```

Comment:
Wait, I'm starting to understand - this is the case where no Sidecar is present, not the Sidecar named "default" or with no label selector ( or whatetever is the current way to indicate that a Sidecar applies to the entire namespace ) ? 

This is extremely confusing  - 'default' should not be 'no sidecar present', but the per-namespace sidecar.
And risky - if a namespace has no sidecars, it should bypass all this new code so we get guaranteed previous behavior.

Reply:
I dont understand what you are saying. The logic here is straightforward: 
 Find the most specific sidecar for a given namespace. 
 If none found, create a sidecar object on the fly that has exact same semantics as of 4 months ago: a sidecar can reach any other service in any namespace.

If the user wants to get scoping behavior, she has to add a sidecar Config to the namespace. There is no other short cut.

Reply:
2 problems:
1. How is the 'default' sidecar implemented here ? Default is a sidecar without labels, or with the name default - I don't know how it ended up, but we do need a Sidecar that defines egress for all workloads. We can't require users to create one sidecar for each resource with the same egress.

2. If no Sidecar exists in the namespace - we should just return nil and keep using the current code ( i.e. return the full list of services/etc), to de-risk the existing code. There is no need to create this fake sidecar and go trough the new Sidecar code instead of using the stable/tested code we had. Both complexity and it's dead code.

Reply:
Per recent changes, while we always return a default sidecarWrapper object [where we store some cached output], all the other code checks if this sidecar Has a config object or not. If not, it takes the old path. Else the newer path.